### DISCUSSION TITLE
【Lecture 3】Circom IsZero

### DISCUSSION THREAD

**Discussion by Demian101** [zkp-co-learning/zkp-co-learn#117] [Q&A（提问题在隔壁~）] [answered]
```circom
ab <-- a*b
ab === a*b
c <-- ab * ab
c === ab * ab
```

读 [这篇文章](https://dev.to/spalladino/a-beginners-intro-to-coding-zero-knowledge-proofs-c56) 看到一个说法：

....  `ab === a*b` means add a constraint that forces ab to be equal to a*b, which gets ignored during execution. 

他说执行期间（execution）`ab === a*b` 会被忽略？

```rust
template IsZero () {
    signal input in;
    signal output out;
    signal inv;
    
    inv <−− in!=0 ? 1/in : 0;  // 赋值, inv 即 inverse, 倒数
    out <-- -in*inv +1;        // 赋值
    out === −in*inv +1; // 电路约束 gate0, gate1, gate2
    in*out === 0;       // 电路约束 gate3
}
```

讲道理，在 IsZero template 中，如果 `===` 被忽略，那么电路岂不是很容易被攻击？

所以 `===` 约束应该在执行期也需要被满足吧？

也就是说，`===` 约束 不仅会在编译器被编译成电路结构，也会在执行期被拿来运行验证？

**Comment by readygo67** [comment, marked as answer]
"==="约束 不仅会在编译器被编译成电路结构，也会在执行期被拿来运行验证。

<img width="1233" alt="image" src="https://github.com/zkp-co-learning/zkp-co-learn/assets/78890754/53b43e9d-13bd-45ce-b2f0-8963e682aeaf">
